### PR TITLE
fix: avoid concealing build errors

### LINKED ISSUE #5413: Improper import in route does not throw but builds nothing
### Describe the bug

When a page (a .svelte-file in the src/routes directory) has a broken import: 
I run `npm run build`.
No errors are thrown, no warnings are given and the site doesnt build. 

Expected behaviour would be one of: 

* Build fails with error
* Build succeeds with warning
* Build succeeds without warning, but throws in runtime

### Reproduction

https://github.com/iverks/sveltekit_import_bug

`npm install`
`npm run build` - nothing gets built, but no errors

Then you can comment out the import in `index.svelte`, and it builds just fine

### Logs

```shell
PS C:\Users\ivism\Progging\bug> npm run build
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.

> bug@0.0.1 build
> vite build

vite v2.9.13 building for production...
✓ 13 modules transformed.
```

### System Info

```shell
Testen on Windows 10 and windows 11.

From win10:
  System:
    OS: Windows 10 10.0.19042
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
    Memory: 4.37 GB / 15.71 GB
  Binaries:
    Node: 16.16.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 8.5.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.423.0), Chromium (103.0.1264.44)
    Internet Explorer: 11.0.19041.1
```

### Severity

serious, but I can work around it

### Additional Information

About severity: I guess it could be considered "annoying", but its terrible DX and was almost impossible to debug.
Tested with adapter-node and adapter-static.
This is using the latest package.json, vite.config.js, svelte.config.js setup  supplied by `npm create svelte app-name`

### PR DESCRIPTION
**Pull request by gtm-nayan**
closes #5413 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

File: packages/kit/src/vite/index.js [hunk 1]
```diff
@@ -99,6 +99,11 @@ function kit() {
 	 */
 	let paths;
 
+	/**
+	 * @type {Error | undefined}
+	 */
+	let build_error;
```
Comment (Rich-Harris) [inline packages/kit/src/vite/index.js:105]:
```suggestion
	/** @type {Error | undefined} */
	let build_error;
```

**Review by Rich-Harris — APPROVED** [review]

**Review by bluwy — APPROVED** [review]

**Comment by Rich-Harris** [comment]
This feels like quite a bit of complexity. IIRC the reason we're not doing the adapt step inside `writeBundle` is that calling `process.exit(0)` in that hook risks interfering with other plugins. What if we just followed through on #5306 — could we avoid all this?

**Comment by benmccann** [comment]
I had wondered as well and https://github.com/sveltejs/kit/issues/5306 would certainly be nice to fix given the number of issues it has caused, but I think the answer is that we still need this because what would it mean to run the adapter on a failed build? I talked with patak that we need additional hooks at the end of the Vite build, so I think that may be part of the answer long-term as well

**Comment by Rich-Harris** [comment]
If adapting was part of `writeBundle` then it _wouldn't_ run on a failed build. we wouldn't get that far into the function

**Comment by benmccann** [comment]
Yes, but it would break Vite plugins that output anything during `writeBundle` as the adapter could no longer utilize their output

**Comment by Rich-Harris** [comment]
Ah, I see. What does that mean in practice though? Are there any adapters that depend on the output of a Vite plugin? Even if so, it feels like all that would need to happen is that the SvelteKit plugin goes after any plugins that generate output the adapter depends on.

**Comment by benmccann** [comment]
> Ah, I see. What does that mean in practice though? Are there any adapters that depend on the output of a Vite plugin?

It's not really specific to any adapter. E.g. `vite-plugin-pwa` generates files during `writeBundle`, but they weren't making them from `.svelte-kit` to `build` or wherever until we moved the adapt phase to happen later

> Even if so, it feels like all that would need to happen is that the SvelteKit plugin goes after any plugins that generate output the adapter depends on.

Unfortunately not because all `writeBundle` hooks run in parallel

File: packages/kit/src/vite/index.js [hunk 2]
```diff
@@ -109,6 +109,12 @@ function kit() {
 	 */
 	let paths;
 
+	/** @type {Error | undefined} */
+	let build_error;
+
+	/** @type {Error | undefined} */
+	let write_bundle_error;
```
Comment (Rich-Harris) [inline packages/kit/src/vite/index.js:116]:
```suggestion
	let errored = false;
```

File: packages/kit/src/vite/index.js [hunk 3]
```diff
@@ -247,6 +253,9 @@ function kit() {
 		 * Clears the output directories.
 		 */
 		buildStart() {
+			// Reset errors in watch mode
+			build_error = write_bundle_error = undefined;
```
Comment (Rich-Harris) [inline packages/kit/src/vite/index.js:257]:
```suggestion
			errored = false;
```

File: packages/kit/src/vite/index.js [hunk 4]
```diff
@@ -256,6 +265,10 @@ function kit() {
 			}
 		},
 
+		buildEnd(err) {
+			build_error = err;
```
Comment (Rich-Harris) [inline packages/kit/src/vite/index.js:269]:
```suggestion
			errored = !!err;
```

File: packages/kit/src/vite/index.js [hunk 5]
```diff
@@ -266,80 +279,86 @@ function kit() {
...
-				if (svelte_config.kit.paths.assets) {
-					throw new Error('Cannot use service worker alongside config.kit.paths.assets');
+					await build_service_worker(options, prerendered, client.vite_manifest);
 				}
-
-				log.info('Building service worker');
-
-				await build_service_worker(options, prerendered, client.vite_manifest);
+			} catch (error) {
+				throw (write_bundle_error = /** @type {Error} */ (error));
```
Comment (Rich-Harris) [inline packages/kit/src/vite/index.js:361]:
```suggestion
				errored = true;
				throw (error);
```

File: packages/kit/src/vite/index.js [hunk 6]
```diff
@@ -354,6 +373,11 @@ function kit() {
 			if (!is_build) {
 				return; // vite calls closeBundle when dev-server restarts, ignore that
 			}
+
+			if (build_error || write_bundle_error) {
```
Comment (Rich-Harris) [inline packages/kit/src/vite/index.js:377]:
```suggestion
			if (errored) {
```

**Review by Rich-Harris — COMMENTED** [review]
given that we're not using the actual values of `build_error` or `write_bundle_error` anywhere, can we just track it as a boolean? suggestions inline

**Comment by Rich-Harris** [comment]
in fact, could we make it even simpler than my suggestion above? rather than tracking failure, we could track success. presumably `writeBundle` isn't called if the build fails, so we could get rid of the `buildEnd` hook, and just set `completed_build = true` at the end of `writeBundle`. opening a PR against this branch, one sec

**Comment by Rich-Harris** [comment]
https://github.com/sveltejs/kit/pull/5536

**Comment by Rich-Harris** [comment]
closing since #5536 was merged — thanks all

**Pull request closed without merging**